operator/apis/mlops/v1alpha1/pipeline_types.go

```diff
@@ -25,42 +25,77 @@ import (
 	"github.com/seldonio/seldon-core/apis/go/v2/mlops/scheduler"
 )
 
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:shortName=mlp
+
+// Pipeline is the Schema for the pipelines API
+type Pipeline struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   PipelineSpec   `json:"spec,omitempty"`
+	Status PipelineStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// PipelineList contains a list of Pipeline
+type PipelineList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []Pipeline `json:"items"`
+}
+
 // PipelineSpec defines the desired state of Pipeline
 type PipelineSpec struct {
 	// External inputs to this pipeline, optional
 	Input *PipelineInput `json:"input,omitempty"`
+
 	// The steps of this inference graph pipeline
 	Steps []PipelineStep `json:"steps"`
+
 	// Synchronous output from this pipeline, optional
 	Output *PipelineOutput `json:"output,omitempty"`
 }
 
+// +kubebuilder:validation:Enum=inner;outer;any
 type JoinType string
 
 const (
+	// data must be available from all inputs
 	JoinTypeInner JoinType = "inner"
+	// data will include any data from any inputs at end of window
 	JoinTypeOuter JoinType = "outer"
-	JoinTypeAny   JoinType = "any"
+	// first data input that arrives will be forwarded
+	JoinTypeAny JoinType = "any"
 )
 
 type PipelineStep struct {
 	// Name of the step
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
```

Comment:
why is the max 253. Probably makes sense but as an internal string does it need to be?

Reply:
I think you're right that it may not need to be!

My understanding last week when I first went through this was that there was some way that the spec field could end up being used somewhere, where k8s would be unhappy with the length.

That was most likely just the pipeline name, and not a pipeline step name though! I'll revert the validations on this field to avoid being overly strict!

Reply:
✅ Done!